spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessageHeaderPropagation.java

```diff
@@ -85,7 +85,9 @@ private void doPut(MessageHeaderAccessor accessor, String key, String value) {
 						nativeHeaders = new LinkedMultiValueMap<>());
 			}
 			if (nativeHeaders instanceof Map<?, ?>) {
-				((Map) nativeHeaders).put(key, Collections.singletonList(value));
+				Map<String, List<String>> copy = new LinkedMultiValueMap<>((Map<String, List<String>>) nativeHeaders);
```

Comment:
You need to have an `else` for that `if (nativeHeaders == null) {`.
Otherwise you are going to wrap a new `LinkedMultiValueMap()` into one more, almost for nothing.